src/Compiler/lex.fsl

```diff
@@ -1427,14 +1527,137 @@ and tripleQuoteString sargs skip = parse
+      else
+          tripleQuoteString sargs skip lexbuf }
+
+and extendedInterpolatedString sargs skip = parse
+ |  '"' '"' '"'
+    { let (buf, fin, _m, kind, args) = sargs
+      args.interpolationDelimiterLength <- 0
+      let cont = LexCont.Token(args.ifdefStack, args.stringNest)
+      fin.Finish buf kind LexerStringFinisherContext.TripleQuote cont }
+
+ | newline
+    { let (buf, _fin, m, kind, args) = sargs
+      newline lexbuf
+      addUnicodeString buf (lexeme lexbuf)
+      if not skip then
+          STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.ExtendedInterpolated, kind, args.interpolationDelimiterLength, m))
+      else
+          extendedInterpolatedString sargs skip lexbuf }
+
+// The rest is to break into pieces to allow double-click-on-word and other such things
+ | ident
+ | integer
+ | xinteger
+ | anywhite +
+    { let (buf, _fin, m, kind, args) = sargs
+      addUnicodeString buf (lexeme lexbuf)
+      if not skip then
+          STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.ExtendedInterpolated, kind, args.interpolationDelimiterLength, m))
+      else
+          extendedInterpolatedString sargs skip lexbuf }
+
+ | "%" +
+    { let (buf, _fin, m, kind, args) = sargs
+      let numPercents = lexeme lexbuf |> String.length
+      let result() =
+          if not skip then
+              STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.ExtendedInterpolated, kind, args.interpolationDelimiterLength, m))
+          else
+              extendedInterpolatedString sargs skip lexbuf
+      // interpolationDelimiterLength is number of $ chars prepended to opening quotes
+      // If number of consecutive % chars in content is equal to interpolationDelimiterLength,
+      // then that sequence is treated as a format specifier,
+      // as in $"""%3d{42}""" or (equivalent) $$"""%%3d{{42}}""".
+      // Any extra % chars up to interpolationDelimiterLength, are treated simply as regular string content.
+      // 2x interpolationDelimiterLength or more % chars in a sequence will result in an error.
+      let maxPercents = 2 * args.interpolationDelimiterLength - 1
+      if numPercents > maxPercents then
+          let m2 = lexbuf.LexemeRange
+          let rest = result()
+          errorR(Error(FSComp.SR.lexTooManyPercentsInTripleQuote(), m2))
+          rest
+      else
+          // Add two % chars for each % that is supposed to be treated as regular string content
+          // + 1 for a format specifier.
+          let percentsToEmit =
+              if numPercents < args.interpolationDelimiterLength then 2 * numPercents
+              else 2 * (numPercents - args.interpolationDelimiterLength) + 1
+          let s = String.replicate percentsToEmit "%"
+          addUnicodeString buf s
+          result() }
+
+ | "{" +
+    { let (buf, fin, m, kind, args) = sargs
```

Comment:
Can anything be refactored out from here? That would be beneficial for testing, comprehension and VS support (intellisense, GTD).

Reply:
Looked into it, but couldn't come up with a solution that would make it worth - seems to me like it would require something a la defunctionalization to avoid a circular dependency (because you would be factoring out a part of a recursive function)

Reply:
Alright, let it be then.